feat: add Electron app support – tested on macOS M4

Wendy-Nam opened the pull request:
Hi @pranavp10,

This PR adds full support for running the app as an Electron app (both dev and production builds).  
FFmpeg now loads correctly in all environments, and I've verified the full lifecycle: install → quit → relaunch via the Dock on macOS (M4, Apple Silicon).

Tested on macOS only for now, as I don’t currently have access to a Windows machine.  
However, based on the setup, it should work **`cross-platform`**.  
If needed, I’d be happy to help test or adjust for Windows in a follow-up!

---

### ✅ Summary of changes

1. **Added full Electron support (dev + prod)**
   - FFmpeg loads correctly via `.env.local` (with port 3000 fallback in Electron only)
   - Verified full lifecycle on macOS M4: install → quit → relaunch

2. **Reverted temporary `localhost:3000` fallback in `layout.tsx`**
   - Had briefly added it for testing, but removed to avoid scattered config
   - `.env.local` is now the single source for `NEXT_PUBLIC_URL`

3. **Electron packaging setup**
   - Enabled `output: 'export'` in `next.config.js`
   - Updated `package.json`:
     ```json
     "files": [
       "electron/**/*",
       ".next/**/*",
       "node_modules/**/*",
       "package.json",
       "next.config.*"
     ],
     "asar": false
     ```
   - Current macOS build size is ~632MB  
     Likely due to broad file inclusion above — will explore optimizing this in a future PR

4. **DevTools toggle (optional)**
   - In `electron/main.js`, you can uncomment line 58 to enable DevTools:
     ```js
     // if (!isProd) win.webContents.openDevTools();
     ```

---

### 🎥 Preview videos (unlisted)

1. [Electron dev build – FFmpeg and UI working (macOS)](https://www.youtube.com/watch?v=unzpJlfn1EQ)
2. [Electron prod build – Full lifecycle test on macOS M4 (Apple Silicon)](https://www.youtube.com/watch?v=xHDLTgpkuKc)
3. [Web dev build – Port mismatch (`3001` vs `3000`) workaround via `.env.local`](https://www.youtube.com/watch?v=Ra5TMhDJblA)

---

Let me know if you'd prefer a smaller scoped PR or any specific adjustments.  
Thanks again — love what you’ve built 🙌

Really impressed that it works fully offline and compresses so well — super thoughtful work.  
If the processing speed ever gets improved a bit, it might make the experience even better. Just a thought!

Wendy-Nam committed:
Add Electron macOS app support & .env fallback for metadataBase

- Add Electron build for macOS (Next.js server runs in background, offline support)
- Fallback to http://localhost:3000 for metadataBase if NEXT_PUBLIC_URL is not set
- Note: Safari (ARM64/M4) does not work with local build, tested with sample mp4
- Added comments for .env usage and fallback

Wendy-Nam committed:
docs: clarify .env.local setup step in README

Wendy-Nam committed:
feat: support Electron (dev + prod) with FFmpeg loading fix

- Fix FFmpeg path in packaged builds
- Remove inline URL fallback (use .env only)
- Enable export mode in next.config.js
- Update Electron build scripts
- Add optional DevTools toggle in main.js
- Verified full flow on macOS (install, run, relaunch)

pranavp10 commented:
@Wendy-Nam, thanks so much for all your help! I’ll test it on my system. We should add a download button on the home page and try to provide separate downloads for macOS and Windows —  it’ll be really helpful for all users. 

Also, would it be possible to add a GitHub Action to automatically create builds for mac and windows? That would streamline the process a lot.

Wendy-Nam commented:
If it’s helpful, I can share a pre-built macOS `.dmg` (built on Apple Silicon, ARM64) — for reference, testing, or lightweight distribution.

Since macOS builds require macOS itself, I understand it might not be convenient to generate locally.

Also, if this version ever feels suitable, it might be possible to prepare it for the Mac App Store in the future.  
(Of course, that would require an Apple Developer account and a $99/year membership — just a thought!)

Wendy-Nam commented:
> @Wendy-Nam, thanks so much for all your help! I’ll test it on my system. We should add a download button on the home page and try to provide separate downloads for macOS and Windows — it’ll be really helpful for all users.
> 
> Also, would it be possible to add a GitHub Action to automatically create builds for mac and windows? That would streamline the process a lot.

@pranavp10 

This Git workflow suggestion sounds great — and aligns well with my interest in infra and automation!

I'd love to help set up GitHub Actions for macOS and Windows builds.  
That said, I'm currently preparing for a time-sensitive interview, so I may not be able to contribute immediately.  
But once I’m available, I’d be happy to jump in — and I believe it won’t take too long.

In the meantime, here’s a **pre-release macOS `.dmg` build** you can check out (for distribution or reference):  
🔗 [Google Drive Link](https://drive.google.com/drive/folders/1GO_V90CnAMldLDIsSUHoKqQaR3y5cxd-?usp=share_link)

For Windows, we’ll likely need a separate build setup or a contributor with a Windows environment.

Wendy-Nam commented:
@pranavp10 

I'm a bit tied up with a pre-interview assignment for the next couple of days, but I’ll try looking into the workflow setup later this week.  
Electron CI/CD for macOS seems tricky with GitHub Actions, so I’m exploring alternatives — including tools like Bitrise.  

Found this helpful article too, will dig in when I get a chance:  

🔗 https://github.com/marketplace/actions/electron-builder-action
🔗 https://medium.com/@charlesugbana04/building-a-ci-cd-pipeline-for-publishing-an-electron-desktop-application-across-three-os-platforms-7f68a1a1ab53

pranavp10 commented on package-lock.json:
We don't need This file

Wendy-Nam replied:
Okay that's fine for me. feel free to erase.

pranavp10 commented on package.json:
```diff
@@ -25,11 +25,33 @@
+    ".next"
+  ],
+  "build": {
+    "appId": "com.yourdomain.videocompress",
```
change this to sh.prolab.videocompress

pranavp10 reviewed (commented):
Rest looks good

Wendy-Nam committed:
feat: implement offline Electron app with filesystem-based architecture

- Remove Next.js server dependency to prevent port conflicts and ensure complete independence
- Fix production build issues (empty screen) with optimized build settings
- Reduce package size from 500MB+ to ~120MB through selective file inclusion
- Add FFmpeg WASM loading via IPC handlers for offline video processing
- Enable cross-platform desktop functionality without external dependencies

Wendy-Nam committed:
chore: add GitHub Actions workflows for Electron CI/CD

Wendy-Nam committed:
chore: add .env creation step to workflow

Wendy-Nam committed:
chore: split CI workflow by OS, run macOS last

Wendy-Nam committed:
chore: split CD workflow, fix icon resolution issue

- split CD workflow for macOS runner
- switch release build scripts from npx to Bun and direct electron-builder commands
- update win icon to 256x256+ (.ico)
- update linux/mac icon to 512x512 (.icns)

Wendy-Nam committed:
docs: add CI/CD and release guide for offline desktop app

Wendy-Nam committed:
chore: release 0.1.1 – desktop app added

- add offline desktop app support and CI/CD workflow
- add release build targets: `.dmg`, `.zip`, `.exe`, `.AppImage`, `.deb`

Wendy-Nam committed:
chore: enable full parallel build for all platforms in build-release workflow

Wendy-Nam commented:
Hi! Hope you’re doing well.
I am sorry for the late follow-up.

I’ve refactored the Electron app architecture and introduced the CI/CD workflows. 
All builds (Linux, Windows, Mac) have been tested and are working as expected.

## Changes

1. **Reduced File Size**
   - Packaging size reduced from ~512MB to 100–190MB.

2. **Fully Serverless Desktop App (Offline Support)**
   - Removed the `Next.js` server background process.
   - Now serves static files using `filesystem APIs`—no server, no ports, no network dependency.
   - The app is now fully offline and serverless.

3. **Automated Multi-OS Releases**
   - Automated builds and releases for Linux, Windows, and Mac via GitHub Actions.
   - Updated documentation for Electron and CI/CD in the docs directory.

**Note:**  
- On Mac, there may be a slight delay when navigating to the video compression page in production builds. This delay seems related to animation libraries or static rendering differences.
- I’ve also found a way to make compression up to 6x faster, but I’ll leave that for a separate PR to keep this one focused. I’ll open an issue for discussion!

Let me know if you have any questions or feedback!

Wendy-Nam commented:
https://github.com/Wendy-Nam/video-compress/releases

My GitHub Actions run uploads all release assets but still fails on the last step, which is not really having actual loss for now. Could be a mismatch between the workflow’s expected file names/formats and what i actually generated, or maybe I've tripped GitHub’s rate limits/cache after too many runs today.

Let’s open an issue on the repo, document these observations, and iterate on a fix.
